neuralpp/test/quick_tests/symoblic/z3_usage_test.py

```diff
@@ -53,13 +52,47 @@ def test_z3_solve_nonlinear_polynomial() -> None:
     s = Solver()
     s.add(x ** 2 + y ** 2 > 3, x ** 3 + y < 5)
     assert s.check() == sat
-    # call s.model() to get a solution
+    # one can call always s.model() to get a solution if check() == sat
 
 
-# def test_bitvec() -> None:
-#
-#
-# def test_function() -> None:
-#
-#
-# def test_quantifier() -> None:
+def test_bitvec() -> None:
+    """ BitVec is Z3's term for bit-vector. For example, 16-bit integer is a bit-vector. """
+    a = BitVecVal(-1, 16)
+    b = BitVecVal(65535, 16)
+    assert is_valid(a == b)  # -1 (signed) is 65535 (unsigned) in 16-bit representation.
+    a = BitVecVal(-1, 32)
+    b = BitVecVal(65535, 32)
+    assert is_valid(a != b)  # -1 is not 65535 in 32-bit representation.
+
+
+def test_function() -> None:
+    """ In Z3, functions are uninterpreted and total. Uninterpreted means it's just a name, we cannot give it
+    any definition or interpretation; total means it has no side effects, like functions in functional language.
+    """
+    x = Int('x')
+    f = Function('f', IntSort(), IntSort())
+    assert is_valid(Implies(f(x) == x, f(f(f(x))) == x))
+
+
+def test_quantifier() -> None:
+    """ Z3 also supports quantifiers, such as `forall`, `exists`. """
+    x, y = Ints('x y')
+    f = Function('f', IntSort(), IntSort(), IntSort())
+    assert is_valid(Implies(ForAll([x, y], f(x, y) >= x), f(0, 5) >= 0))  # if forall x y, f(x,y)>=x, then f(0,5)>=0
+    assert is_valid(Implies(Exists([x], f(x, y) == x), Not(ForAll([x], f(x, y) != x))))  # exists
+
+
+def test_sum() -> None:
+    """
+    `Sum` is not a quantifier in Z3, it's just a Python function.
```

Comment:
I don't think it is accurate to say it is a _Python_ function. If we use `Sum` on Z3 variables, it is not going to solve it with the Python operator +, but will need to actually symbolically solve that. I would replace the word "Python" here by "interpreted".

Reply:
Yes that makes sense.